### ISSUE TITLE
False positive InvalidReturnStatement and InvalidReturnType due to array shape not equal array

### ISSUE THREAD

**Issue by kkmuffme** [vimeo/psalm#9427]
https://psalm.dev/r/e03450ce4c

`array{a: false, b: false, c: false}` and `non-empty-array<'a'|'b'|'c', false>` are identical and no error should be given

**Comment by orklah** [comment]
I don't agree that they're identical `['a', false]` is a valid `non-empty-array<'a'|'b'|'c', false>` but is not a valid `array{a: false, b: false, c: false}`

This snippet still has an issue because in fact, the foreach should produce a `array{a: false, b: false, c: false}` if it was inferred optimally

**Comment by orklah** [comment]
indeed!
